frontend/src/tests/import-export.test.js

```diff
@@ -2,47 +2,171 @@
-            last_name: "Smith",
-            email: "a@b.com",
-            utorid: "food",
-        },
-        {
-            first_name: "Grandpa",
-            last_name: "Boobie",
-            email: "a@d.com",
-            utorid: "fooc",
-        },
-    ];
+    /**
+     *  Create a `File` object containing of the specified format.
+     * This function is a copy of function dataToFile in ../libs/importExportUtils.
+     * The only difference is this function returns a FileAPI File object but not a File object.
+     * The reason is Jest is running on NodeJS which does not support File object, and the file-apis File object has a different constructor with File.
+     * By calling this function only for testing, the app's export functionality will not be affected,
+     * while the test can still test the export functionality.
+     *
+     * @param {{toSpreadsheet: func, toJson: func}} formatters - Formatters return an array of objects (usable as spreadsheet rows) or a javascript object to be passed to JSON.stringify
+     * @param {"xlsx" | "csv" | "json"} dataFormat
+     * @param {string} filePrefix
+     * @returns {FileAPI.File}
+     */
+    function dataToFileForTest(formatters, dataFormat, filePrefix = "") {
+        const fileName = `${filePrefix}${
+            filePrefix ? "_" : ""
+        }export_${new Date().toLocaleDateString("en-CA", {
+            year: "numeric",
+            month: "numeric",
+            day: "numeric",
+        })}`;
+
+        if (dataFormat === "spreadsheet" || dataFormat === "csv") {
+            const workbook = XLSX.utils.book_new();
+            const sheet = XLSX.utils.aoa_to_sheet(formatters.toSpreadsheet());
+            XLSX.utils.book_append_sheet(workbook, sheet, "Instructors");
+
+            const bookType = dataFormat === "csv" ? "csv" : "xlsx";
+
+            // We convert the data into a blob and return it so that it can be downloaded
+            // by the user's browser
+            const file = new File({
+                buffer: [XLSX.write(workbook, { type: "array", bookType })],
+                name: `${fileName}.${bookType}`,
+                type:
+                    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+            });
+            return file;
+        }
+
+        if (dataFormat === "json") {
+            const file = new File({
+                buffer: [JSON.stringify(formatters.toJson(), null, 4)],
+                name: `${fileName}.json`,
+                type: "application/json",
+            });
+            return file;
+        }
+
+        throw new Error(
+            `Cannot process data to format "${dataFormat}"; try "spreadsheet" or "json".`
+        );
+    }
+
+    /**
+     * Construct a FileAPI `File` object based on pre-defined object data and input file type
+     *
+     * @param {"xlsx" | "csv" | "json"} dataFormat
+     * @returns {FileAPI.File}
+     */
+    function getInstructorsDataFile(dataFormat) {
```

Comment:
Isn't this another redefinition of an existing function? Can you export it from where it's originally defined?

Reply:
Yes it is, but the original one ("prepareData" in \frontend\src\views\instructors\import-export.js) is inside another function, and I can not export it because it's not at top level.